# Overview

In this project, I will take the role of a data scientist to develop and evalute machine learning models that predict the price of used cars. I will divide the project into **three main** steps:

## STEP 1: Data Preprocessing

In this step, we will:

- Handle missing data by imputing values
- Address duplicate entries in the dataset.
- Encode categorical features using one-hot encoding.
- Normalize numerical features to ensure consistent scales.

## STEP 2: EDA (Exploratory Data Analysis)

In this step, we will:

- Visualize data to understand feature distributions, correlations, and outliers.
- Detect and manage outliers. 
- Explore relationships between features and the target variable (car prices).
- Identify key insights and patterns in the data that can inform model selection and feature engineering.
- Gain a deeper understanding of the dataset to guide model development.

## STEP 3: Model Development and Evaluation

In this step, we will:

- Select appropriate machine learning algorithms (e.g., regression models) for prediction.
- Perform feature engineering to create relevant features.
- Train and tune models using the training dataset.
- Evaluate model performance using metrics such as Mean Squared Error (MSE), Mean Absolute Error (MAE), and R-squared (R²).
- Validate models on the validation dataset and fine-tune hyperparameters.
- Select the best-performing model and assess its performance on the test dataset.
- Present the final model for predicting used car prices.


# STEP 1: Data Preprocessing 

<html>
<head>

</head>
<body>

<h3>Data Preprocessing Steps Overview</h3>

<ol>
  <li><strong>Load the Data</strong>:
    <ul>
      <li>Loaded the training and test datasets.</li>
    </ul>
  </li>
  <li><strong>Handle Missing Values</strong>:
    <ul>
      <li>Identified missing values in the engine_capacity feature.</li>
      <li>Imputed missing values in engine_capacity with the median value for both training and test datasets.</li>
    </ul>
  </li>
  <li><strong>Convert Data Types</strong>:
    <ul>
      <li>Converted categorical features to 'category' data type.</li>
      <li>Converted boolean features to integer data type.</li>
    </ul>
  </li>
  <li><strong>Feature Engineering</strong>:
    <ul>
      <li>Calculated the age_of_car feature and dropped the original year_produced feature.</li>
      <li>Dropped the engine_fuel feature after analyzing its relationship with engine_type.</li>
    </ul>
  </li>
  <li><strong>One-Hot Encoding</strong>:
    <ul>
      <li>Applied one-hot encoding to categorical features (transmission, engine_type, ownership, type_of_drive) and dropped the original features.</li>
    </ul>
  </li>
  <li><strong>Save Preprocessed Data</strong>:
    <ul>
      <li>Saved the preprocessed training and test datasets to CSV files for future use.</li>
    </ul>
  </li>
</ol>

</body>
</html>

#### Importing libraries 

In [1]:
import pandas as pd 
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

#### Surpressing warnings

In [2]:
# Surpress warnings:
def warn(*args, **kwargs):
    pass
import warnings
warnings.warn = warn

#### Loading the two datasets

In [3]:
# Load the datasets
data_train = pd.read_csv(r"C:\Users\Shekh\Downloads\CarPrice_Assignment - Copy.csv"')
data_test = pd.read_csv(r"C:\Users\Shekh\Downloads\CarPrice_Assignment - Copy.csv"')

In [4]:
#Displaying the first few rows
data_train.head()

manufacturer_name transmission  color  odometer_value  year_produced  \
0        Volkswagen    automatic  black          130000           2016   
1           Renault       manual  brown          149000           2012   
2               Kia    automatic  brown          110000           2014   
3              Opel    automatic  other          255100           2007   
4             Mazda       manual   blue          650000           1999   

  engine_fuel engine_type  engine_capacity  body_type  has_warranty ownership  \
0      diesel      diesel              1.6  universal         False     owned   
1    gasoline    gasoline              1.6      sedan         False     owned   
2    gasoline    gasoline              1.6  hatchback         False     owned   
3    gasoline    gasoline              1.8  hatchback         False     owned   
4    gasoline    gasoline              2.0      sedan         False     owned   

  type_of_drive  is_exchangeable  number_of_photos  number_of_maintenance  \
0         front             True                17                     38   
1         front            False                 9                      3   
2         front            False                 5                     10   
3         front            False                10                      4   
4         front             True                 5                      7   

   duration_listed  price_usd  
0               67    13150.0  
1              100     7500.0  
2               91    12200.0  
3               91     4950.0  
4               62     3000.0

In [5]:
# Displying details about the dataframe
data_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 17 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   manufacturer_name      50000 non-null  object 
 1   transmission           50000 non-null  object 
 2   color                  50000 non-null  object 
 3   odometer_value         50000 non-null  int64  
 4   year_produced          50000 non-null  int64  
 5   engine_fuel            50000 non-null  object 
 6   engine_type            50000 non-null  object 
 7   engine_capacity        49985 non-null  float64
 8   body_type              50000 non-null  object 
 9   has_warranty           50000 non-null  bool   
 10  ownership              50000 non-null  object 
 11  type_of_drive          50000 non-null  object 
 12  is_exchangeable        50000 non-null  bool   
 13  number_of_photos       50000 non-null  int64  
 14  number_of_maintenance  50000 non-null  int64  
 15  du

In [6]:
data_train.describe()

odometer_value  year_produced  engine_capacity  number_of_photos  \
count    50000.000000   50000.000000     49985.000000      50000.000000   
mean    249219.881980    2002.945620         2.056894          9.662420   
std     136361.343789       8.067141         0.675106          6.145263   
min          0.000000    1942.000000         0.200000          1.000000   
25%     158000.000000    1998.000000         1.600000          5.000000   
50%     250000.000000    2003.000000         2.000000          8.000000   
75%     325000.000000    2009.000000         2.300000         12.000000   
max    1000000.000000    2019.000000         8.000000         86.000000   

       number_of_maintenance  duration_listed     price_usd  
count           50000.000000     50000.000000  50000.000000  
mean               16.256940        80.739440   6654.891274  
std                41.537349       113.606476   6467.096697  
min                 1.000000         0.000000      1.000000  
25%                 2.000000        23.000000   2100.000000  
50%                 5.000000        59.000000   4800.000000  
75%                16.000000        91.000000   8990.000000  
max              1273.000000      2232.000000  50000.000000

In [7]:
data_train.shape

(50000, 17)

**Feature Description**
<body>
<p><strong>Odometer value:</strong> Shows a wide range of values, suggesting that cars in various conditions are present in the dataset.</p>
<p><strong>Year produced:</strong> The cars in the dataset range from vintage models (1937) to recent ones (2019), indicating a broad spectrum of car ages.</p>
<p><strong>Engine capacity:</strong> There is significant variation in engine capacity, indicating a mix of different car types.</p>
<p><strong>Price (USD):</strong> As the potential target variable, it has a wide range, which is typical for used car prices and indicates variability in car values.</p>
</body>
</html>



## Missing values

In [8]:
data_train.isnull().sum(), data_test.isnull().sum()

(manufacturer_name         0
 transmission              0
 color                     0
 odometer_value            0
 year_produced             0
 engine_fuel               0
 engine_type               0
 engine_capacity          15
 body_type                 0
 has_warranty              0
 ownership                 0
 type_of_drive             0
 is_exchangeable           0
 number_of_photos          0
 number_of_maintenance     0
 duration_listed           0
 price_usd                 0
 dtype: int64,
 manufacturer_name        0
 transmission             0
 color                    0
 odometer_value           0
 year_produced            0
 engine_fuel              0
 engine_type              0
 engine_capacity          5
 body_type                0
 has_warranty             0
 ownership                0
 type_of_drive            0
 is_exchangeable          0
 number_of_photos         0
 number_of_maintenance    0
 duration_listed          0
 price_usd                0
 dtype: int64)

<html>
<head>

<style>
table td, th {
  text-align: left;
}
</style>
</head>
<body>

<h3>Missing Values in Training and Test Data</h3>

<table style="width:100%">
  <tr>
    <th style="text-align: left;">Dataset</th>
    <th style="text-align: left;">Feature</th>
    <th style="text-align: left;">Number of Missing Values</th>
  </tr>
  <tr>
    <td style="text-align: left;">Training Data</td>
    <td style="text-align: left;">engine_capacity</td>
    <td style="text-align: left;">15</td>
  </tr>
  <tr>
    <td style="text-align: left;">Test Data</td>
    <td style="text-align: left;">engine_capacity</td>
    <td style="text-align: left;">5</td>
  </tr>
  <tr>
    <td style="text-align: left;">Both Datasets</td>
    <td style="text-align: left;">All other features</td>
    <td style="text-align: left;">0</td>
  </tr>
</table>

<p>Next Steps</p>

<ul>
  <li>Imputation: Replace the missing values with a specific value, such as the mean, median, or mode of the feature.</li>
  <li>Deletion: Remove the rows with missing values, especially if the number of such rows is relatively small compared to the entire dataset.</li>
  <li>Prediction: Use other features in the dataset to build a model that can predict and fill in the missing values.</li>
</ul>

</body>
</html>

Since the missing values are a very small percentage, I will be using **imputation with the median**. Imputing with the median helps to maintain the original distribution of the data. This is crucial for machine learning models as changes in the distribution can lead to biased or incorrect predictions.

In [9]:
# Handling Missing Values for training dataset

# Calculate the median value of engine_capacity
median_engine_capacity = data_train['engine_capacity'].median()

# Impute missing values in engine_capacity with the median value
data_train['engine_capacity'].fillna(median_engine_capacity, inplace=True)

In [10]:
# Handling Missing Values for testing dataset

# Calculate the median value of engine_capacity
median_engine_capacity = data_test['engine_capacity'].median()

# Impute missing values in engine_capacity with the median value
data_test['engine_capacity'].fillna(median_engine_capacity, inplace=True)

In [11]:
# Check if there are any remaining missing values
has_null = (data_train.isnull().sum().any()) or (data_test.isnull().sum().any())

if not has_null:
    print('There are no more null values')
else:
    print('There are still some null values')

There are no more null values


## Data type Conversion

In [12]:
# Check the data types
data_train.dtypes

manufacturer_name         object
transmission              object
color                     object
odometer_value             int64
year_produced              int64
engine_fuel               object
engine_type               object
engine_capacity          float64
body_type                 object
has_warranty                bool
ownership                 object
type_of_drive             object
is_exchangeable             bool
number_of_photos           int64
number_of_maintenance      int64
duration_listed            int64
price_usd                float64
dtype: object

Converting categorical features to the **'category'** data type in pandas is a great practice, especially when dealing with machine learning models. This conversion can lead to more efficient memory usage and can also enable the use of certain pandas functionalities specific to categorical data.

In [13]:
# Convert categorical features to 'category' data type
categorical_features = ['manufacturer_name', 'transmission', 'color', 'engine_fuel', 'engine_type', 'body_type', 'ownership', 'type_of_drive']
data_train[categorical_features] = data_train[categorical_features].astype('category')
data_test[categorical_features] = data_test[categorical_features].astype('category')

## Feature Engineering

#### Creating a feature 'age_of_car' from 'year_produced' column

In [14]:
current_year = max(data_train['year_produced'].max(),data_test['year_produced'].max())
print(f'We assume the current year is {current_year}')

We assume the current year is 2019


In [15]:
#Create 'age_of_car' feature in place of 'year_produced'
data_train['year_produced'] = current_year - data_train['year_produced']
data_test['year_produced'] = current_year - data_test['year_produced']

# Rename the 'year_produced' column to 'age_of_car' in both DataFrames
data_train = data_train.rename(columns={'year_produced': 'age_of_car'})
data_test = data_test.rename(columns={'year_produced': 'age_of_car'})

# Display the first few rows to verify the new feature
data_train.head()

manufacturer_name transmission  color  odometer_value  age_of_car  \
0        Volkswagen    automatic  black          130000           3   
1           Renault       manual  brown          149000           7   
2               Kia    automatic  brown          110000           5   
3              Opel    automatic  other          255100          12   
4             Mazda       manual   blue          650000          20   

  engine_fuel engine_type  engine_capacity  body_type  has_warranty ownership  \
0      diesel      diesel              1.6  universal         False     owned   
1    gasoline    gasoline              1.6      sedan         False     owned   
2    gasoline    gasoline              1.6  hatchback         False     owned   
3    gasoline    gasoline              1.8  hatchback         False     owned   
4    gasoline    gasoline              2.0      sedan         False     owned   

  type_of_drive  is_exchangeable  number_of_photos  number_of_maintenance  \
0         front             True                17                     38   
1         front            False                 9                      3   
2         front            False                 5                     10   
3         front            False                10                      4   
4         front             True                 5                      7   

   duration_listed  price_usd  
0               67    13150.0  
1              100     7500.0  
2               91    12200.0  
3               91     4950.0  
4               62     3000.0

#### Check Differences Between 'engine_fuel' and 'engine_type'
Investigating the unique combinations of 'engine_fuel' and 'engine_type' to understand their relationship and to decide if one of them can be dropped to reduce redundancy.

In [16]:
# Get unique combinations of 'engine_fuel' and 'engine_type'
unique_combinations = data_train.groupby(['engine_fuel', 'engine_type']).size().reset_index(name='Count')

# Remove rows where 'Count' is zero in the unique_combinations DataFrame
unique_combinations = unique_combinations[unique_combinations['Count'] > 0].reset_index(drop=True)

unique_combinations

engine_fuel engine_type  Count
0         diesel      diesel  16657
1       electric    electric     15
2            gas    gasoline   1753
3       gasoline    gasoline  31255
4  hybrid-diesel      diesel      4
5  hybrid-petrol    gasoline    316

**Investigate 'engine_fuel' and 'engine_type'**.
The unique combinations of 'engine_fuel' and 'engine_type' were examined. Here are some key combinations along with their counts:

- **Diesel Engine:** 16,657 cars
- **Electric Engine:** 15 cars
- **Gasoline Engine:** 31,255 cars
- **Gas Engine:** 1,753 cars
- **Hybrid Diesel:** 4 cars
- **Hybrid Petrol:** 316 cars

In summary, it is safe to drop the engine_fuel column since the combinations are referring to the same feature

In [17]:
#drop engine_fuel
data_train = data_train.drop('engine_fuel', axis = 1)
data_test = data_test.drop('engine_fuel', axis = 1)

In [18]:
# Change bool type columns to int (1 and 0)
data_train[['has_warranty','is_exchangeable']] = data_train[['has_warranty','is_exchangeable']].astype('int')
data_test[['has_warranty','is_exchangeable']] = data_test[['has_warranty','is_exchangeable']].astype('int')

data_train.head()

manufacturer_name transmission  color  odometer_value  age_of_car  \
0        Volkswagen    automatic  black          130000           3   
1           Renault       manual  brown          149000           7   
2               Kia    automatic  brown          110000           5   
3              Opel    automatic  other          255100          12   
4             Mazda       manual   blue          650000          20   

  engine_type  engine_capacity  body_type  has_warranty ownership  \
0      diesel              1.6  universal             0     owned   
1    gasoline              1.6      sedan             0     owned   
2    gasoline              1.6  hatchback             0     owned   
3    gasoline              1.8  hatchback             0     owned   
4    gasoline              2.0      sedan             0     owned   

  type_of_drive  is_exchangeable  number_of_photos  number_of_maintenance  \
0         front                1                17                     38   
1         front                0                 9                      3   
2         front                0                 5                     10   
3         front                0                10                      4   
4         front                1                 5                      7   

   duration_listed  price_usd  
0               67    13150.0  
1              100     7500.0  
2               91    12200.0  
3               91     4950.0  
4               62     3000.0

### One-hot encoding catagorical features

We will create a function that takes in the column name in the dataframe and perform one-hot encoding on it

In [19]:
def one_hot_encode_and_insert(data, column_name, index):
    # One-Hot Encoding for the specified column
    dummy = pd.get_dummies(data[[column_name]], columns=[column_name], drop_first=True).astype(int)

    # Drop the original column in the data DataFrame
    data.drop(columns=[column_name], inplace=True)

    # Insert the one-hot encoded columns into the data DataFrame at the specified index
    for col in dummy.columns:
        data.insert(index, f'{column_name}_{col}', dummy[col])
        index += 1
    data.head()

<html>

<body>

<h3>Explanation of drop_first=True</h3>

<p>When you perform one-hot encoding on a categorical feature, each unique value in the feature is transformed into a new binary column. If a categorical feature has <em>N</em> unique values, you would get <em>N</em> binary columns after one-hot encoding.</p>

<p>Setting <code>drop_first=True</code> in <code>pd.get_dummies()</code> drops the first binary column. This is done for two main reasons:</p>

<ul>
  <li><strong>Multicollinearity Avoidance:</strong> Without dropping the first column, the one-hot encoded columns would be perfectly correlated (i.e., linearly dependent), which can cause issues in certain types of models, especially those that are sensitive to multicollinearity, like linear regression.</li>
  <li><strong>Dimensionality Reduction:</strong> Dropping one of the binary columns reduces the dimensionality of the dataset, leading to simpler models and potentially better generalization to new data.</li>
</ul>

<h3>Explanation of the 'transmission_manual' Column</h3>

<p>After one-hot encoding the <code>'transmission'</code> column, if a car has a manual transmission, the <code>'transmission_manual'</code> column will have a value of 1, and 0 otherwise which indicates it's Automatic transmission. This binary column effectively captures the transmission type of the car.</p>

</body>
</html>

In [20]:
# Defining the columns to encode
encoder_columns = ['transmission', 'engine_type', 'ownership', 'type_of_drive']

In [21]:
# For each column defined above, we will disply the unique values
for col in encoder_columns:
    print(f'The unique values of {col} are: \n {data_train[col].unique().tolist()} \n')

The unique values of transmission are: 
 ['automatic', 'manual'] 

The unique values of engine_type are: 
 ['diesel', 'gasoline', 'electric'] 

The unique values of ownership are: 
 ['owned', 'emergency', 'new'] 

The unique values of type_of_drive are: 
 ['front', 'rear', 'all'] 



We will create a data list so that we can iterate some operations on both the training and testing datasets

In [22]:
# Creating a list to include both the training and testing datasets
data = [data_train, data_test]

In [23]:
# One-Hot Encoding for 'transmission'column 
for i in data:
    one_hot_encode_and_insert(i, 'transmission', 1)

In [24]:
# One-Hot Encoding for 'engine_type'column 
for i in data:
    one_hot_encode_and_insert(i, 'engine_type', 5)

In [25]:
# One-Hot Encoding for 'ownership'column 
for i in data:
    one_hot_encode_and_insert(i, 'ownership', 10)

In [26]:
# One-Hot Encoding for 'type_of_drive'column
for i in data:
    one_hot_encode_and_insert(i, 'type_of_drive', 12)

    
data_train.head()

manufacturer_name  transmission_transmission_manual  color  odometer_value  \
0        Volkswagen                                 0  black          130000   
1           Renault                                 1  brown          149000   
2               Kia                                 0  brown          110000   
3              Opel                                 0  other          255100   
4             Mazda                                 1   blue          650000   

   age_of_car  engine_type_engine_type_electric  \
0           3                                 0   
1           7                                 0   
2           5                                 0   
3          12                                 0   
4          20                                 0   

   engine_type_engine_type_gasoline  engine_capacity  body_type  has_warranty  \
0                                 0              1.6  universal             0   
1                                 1              1.6      sedan             0   
2                                 1              1.6  hatchback             0   
3                                 1              1.8  hatchback             0   
4                                 1              2.0      sedan             0   

   ownership_ownership_new  ownership_ownership_owned  \
0                        0                          1   
1                        0                          1   
2                        0                          1   
3                        0                          1   
4                        0                          1   

   type_of_drive_type_of_drive_front  type_of_drive_type_of_drive_rear  \
0                                  1                                 0   
1                                  1                                 0   
2                                  1                                 0   
3                                  1                                 0   
4                                  1                                 0   

   is_exchangeable  number_of_photos  number_of_maintenance  duration_listed  \
0                1                17                     38               67   
1                0                 9                      3              100   
2                0                 5                     10               91   
3                0                10                      4               91   
4                1                 5                      7               62   

   price_usd  
0    13150.0  
1     7500.0  
2    12200.0  
3     4950.0  
4     3000.0

In [27]:
data_test.head()

manufacturer_name  transmission_transmission_manual   color  odometer_value  \
0               BMW                                 0   white          115000   
1     Mercedes-Benz                                 1   other          500000   
2              Fiat                                 1  silver          210000   
3        Mitsubishi                                 0  violet          294000   
4              Opel                                 0    blue          244000   

   age_of_car  engine_type_engine_type_electric  \
0           7                                 0   
1          20                                 0   
2          17                                 0   
3          19                                 0   
4          21                                 0   

   engine_type_engine_type_gasoline  engine_capacity  body_type  has_warranty  \
0                                 1              4.4      sedan             0   
1                                 0              2.2      sedan             0   
2                                 1              1.2  hatchback             0   
3                                 0              3.2        suv             0   
4                                 1              1.6      sedan             0   

   ownership_ownership_new  ownership_ownership_owned  \
0                        0                          1   
1                        0                          1   
2                        0                          1   
3                        0                          1   
4                        0                          1   

   type_of_drive_type_of_drive_front  type_of_drive_type_of_drive_rear  \
0                                  0                                 0   
1                                  0                                 1   
2                                  1                                 0   
3                                  0                                 0   
4                                  1                                 0   

   is_exchangeable  number_of_photos  number_of_maintenance  duration_listed  \
0                1                32                    104              146   
1                0                 7                      9              147   
2                1                16                      7               27   
3                1                10                      2               48   
4                0                 9                     10              116   

   price_usd  
0    20450.0  
1     2600.0  
2     2900.0  
3     7500.0  
4     2200.0

In [28]:
# Rename columns to reduce redundant characters
for d in data:
    d.rename(columns={'transmission_transmission_manual': 'transmission_manual',
                         'engine_type_engine_type_electric': 'engine_electric',
                         'engine_type_engine_type_gasoline': 'engine_gasoline',
                         'ownership_ownership_new':'ownership_new',
                         'ownership_ownership_owned':'ownership_owned',
                         'type_of_drive_type_of_drive_front':'front_drive',
                         'type_of_drive_type_of_drive_rear':'rear_drive'}, inplace=True)

data_train.head()    

manufacturer_name  transmission_manual  color  odometer_value  age_of_car  \
0        Volkswagen                    0  black          130000           3   
1           Renault                    1  brown          149000           7   
2               Kia                    0  brown          110000           5   
3              Opel                    0  other          255100          12   
4             Mazda                    1   blue          650000          20   

   engine_electric  engine_gasoline  engine_capacity  body_type  has_warranty  \
0                0                0              1.6  universal             0   
1                0                1              1.6      sedan             0   
2                0                1              1.6  hatchback             0   
3                0                1              1.8  hatchback             0   
4                0                1              2.0      sedan             0   

   ownership_new  ownership_owned  front_drive  rear_drive  is_exchangeable  \
0              0                1            1           0                1   
1              0                1            1           0                0   
2              0                1            1           0                0   
3              0                1            1           0                0   
4              0                1            1           0                1   

   number_of_photos  number_of_maintenance  duration_listed  price_usd  
0                17                     38               67    13150.0  
1                 9                      3              100     7500.0  
2                 5                     10               91    12200.0  
3                10                      4               91     4950.0  
4                 5                      7               62     3000.0

In [29]:
# Re_check if there is any NaN in the DataFrame
# data = [data_train, data_test]
for d in data:
    has_nan = d.isnull().any().any()

    # Print the result
    print(has_nan)

False
False


In [30]:
# Re_check the datatype of each column
data_train.dtypes

manufacturer_name        category
transmission_manual         int64
color                    category
odometer_value              int64
age_of_car                  int64
engine_electric             int64
engine_gasoline             int64
engine_capacity           float64
body_type                category
has_warranty                int64
ownership_new               int64
ownership_owned             int64
front_drive                 int64
rear_drive                  int64
is_exchangeable             int64
number_of_photos            int64
number_of_maintenance       int64
duration_listed             int64
price_usd                 float64
dtype: object

In [31]:
print(data_train.shape)

(50000, 19)


### Saving the preprocessed data to be used in the next step: EDA

In [32]:
data_train.to_csv('data_train_1.csv', index=False)
data_test.to_csv('data_test_1.csv', index=False)